docs/standard/serialization/binaryformatter-migration-guide/reading-nrbf-payloads.md

```diff
@@ -0,0 +1,206 @@
+---
+title: "BinaryFormatter Migration Guide: Reading NRBF Payloads"
+description: "Reading the payloads that were created using BinaryFormatter."
+ms.date: 5/31/2024
+no-loc: [BinaryFormatter, Serialization]
+dev_langs:
+  - CSharp
+  - VB
+helpviewer_keywords:
+  - "BinaryFormatter"
+  - "serializing objects"
+  - "serialization"
+  - "objects, serializing"
+---
+
+# Reading NRBF payloads
+
+## NrbfDecoder
+
+`NrbfDecoder` is part of the new [System.Formats.Nrbf](https://www.nuget.org/packages/System.Formats.Nrbf) NuGet package. It targets not only .NET 9, but also older monikers like .NET Standard 2.0 and .NET Framework. Which **makes it possible for everyone who uses a supported version of .NET to migrate away from `BinaryFormatter`**.
+
+`NrbfDecoder` can read any payload that was serialized with `BinaryFormatter` and `FormatterTypeStyle.TypesAlways`. `FormatterTypeStyle.TypesAlways` is the default setting that forces BF to emit full type information and does not require any type loading when reading the payload. Which means that `FormatterTypeStyle.TypesWhenNeeded` is not supported.
+
+Only non-zero indexed arrays and types specific to remoting (which were never ported to .NET (Core)) are **not supported**.
+
+`NrbfDecoder` is following these principles to read from **untrusted input**:
+
+- Treating every input as potentially hostile.
+- No type loading of any kind (to avoid remote code execution).
+- No recursion of any kind (to avoid unbound recursion, stack overflow and denial of service).
+- No buffer pre-allocation based on size provided in the payload, if the payload is too small to contain the promised data (to avoid running out of memory and denial of service).
+- Decoding every part of the input only once (to perform the same amount of work as the potential attacker who created the payload).
+- Using collision-resistant randomized hashing to store records referenced by other records (to avoid running out of memory for dictionary backed by an array which size depends on the number of hash code collisions).
+- Only primitive types can be instantiated in an implicit way. Arrays can be instantiated on demand. Other types are never instantiated.
```

Comment:
I'm not really sure any of this belongs in a doc at this high a level.  For any points here, unless it is specifically related to other text later (including explaining a "why" of some sample code), I advocate deleting them.

Reply:
I see your point, let's wait for feedback from other reviewers.

Reply:
I favor keeping this content here as it accomplishes a few goals:

1. It emphasizes up-front that NrbfDecoder is designed to handle untrusted payloads, outlining _how_ it achieves that
2. It preemptively answers questions about why NrbfDecoder _doesn't_ do things folks might want (expect)
3. It raises awareness that these are dangerous behaviors for reading untrusted input, guiding consumers of NrbfDecoder not to simply wrap it with those capabilities

I will add a couple sentences after the bullets to make that last point explicit though.